### ISSUE TITLE
require and List::MoreUtils seg fault

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8506]
Migrated from [rt.perl.org#39650](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=39650) (status was 'rejected')

Searchable as RT39650$

**Comment by p5pRT** [comment]
### From @dmcbride

#### Created by @dmcbride



Example code&#8203;:

\#\!/usr/bin/perl
use strict;
use warnings;

require List&#8203;::MoreUtils;
\# use List&#8203;::MoreUtils \(\);

my @&#8203;v1 = qw\(4 0 0 3\);
my @&#8203;v2 = qw\(4 0 0 3\);

my $ea = List&#8203;::MoreUtils&#8203;::each\_array\(@&#8203;v1\, @&#8203;v2\);
while \(my \($v\, $t\) = $ea\->\(\)\)
\{
\}
\_\_END\_\_

When running the above\, I get a seg fault\.  Switch out the require to
use\, and the seg fault goes away\.  This has been confirmed on 5\.8\.7 on
XP as well in this thread&#8203;: http&#8203;://www\.perlmonks\.org/index\.pl?node\_id=557894

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.8.8:

Configured by Gentoo at Sun May 28 12:22:25 MDT 2006.

Summary of my perl5 (revision 5 version 8 subversion 8) configuration:
  Platform:
    osname=linux, osvers=2.6.16-gentoo-r7, archname=i686-linux-thread-multi
    uname='linux naboo2 2.6.16-gentoo-r7 #1 smp tue may 16 14:39:36 mdt 2006 i686 intel(r) pentium(r) 4 cpu 3.20ghz gnulinux '
    config_args='-des -Darchname=i686-linux-thread -Dcccdlflags=-fPIC -Dccdlflags=-rdynamic -Dcc=i686-pc-linux-gnu-gcc -Dprefix=/usr -Dvendorprefix=/usr -Dsiteprefix=/usr -Dlocincpth=  -Doptimize=-march=pentium4 -O2 -pipe -Duselargefiles -Dd_semctl_semun -Dscriptdir=/usr/bin -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dinstallman1dir=/usr/share/man/man1 -Dinstallman3dir=/usr/share/man/man3 -Dman1ext=1 -Dman3ext=3pm -Dinc_version_list=5.8.0 5.8.0/i686-linux-thread-multi 5.8.2 5.8.2/i686-linux-thread-multi 5.8.4 5.8.4/i686-linux-thread-multi 5.8.5 5.8.5/i686-linux-thread-multi 5.8.6 5.8.6/i686-linux-thread-multi 5.8.7 5.8.7/i686-linux-thread-multi  -Dcf_by=Gentoo -Ud_csh -Dusenm -Dusethreads -Di_ndbm -Di_gdbm -Di_db'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='i686-pc-linux-gnu-gcc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -fno-strict-aliasing -pipe -Wdeclaration-after-statement -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-march=pentium4 -O2 -pipe',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -fno-strict-aliasing -pipe -Wdeclaration-after-statement'
    ccversion='', gccversion='3.4.5 (Gentoo 3.4.5-r1, ssp-3.4.5-1.0, pie-8.7.9)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='i686-pc-linux-gnu-gcc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lpthread -lnsl -lndbm -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.3.6.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.3.6'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.8:
    /etc/perl
    /usr/lib/perl5/vendor_perl/5.8.8/i686-linux-thread-multi
    /usr/lib/perl5/vendor_perl/5.8.8
    /usr/lib/perl5/vendor_perl
    /usr/lib/perl5/site_perl/5.8.8/i686-linux-thread-multi
    /usr/lib/perl5/site_perl/5.8.8
    /usr/lib/perl5/site_perl/5.8.5
    /usr/lib/perl5/site_perl/5.8.6
    /usr/lib/perl5/site_perl/5.8.7
    /usr/lib/perl5/site_perl/5.8.7/i686-linux-thread-multi
    /usr/lib/perl5/site_perl
    /usr/lib/perl5/5.8.8/i686-linux-thread-multi
    /usr/lib/perl5/5.8.8
    /usr/local/lib/site_perl
    .


Environment for perl v5.8.8:
    HOME=/home/dmcbride
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/bin:/bin:/opt/bin:/usr/i686-pc-linux-gnu/gcc-bin/3.4.5:/usr/kde/3.5/bin:/usr/qt/3/bin:/usr/kde/3.4/bin:/share/cvs/bin:/home/dmcbride/bin:/usr/local/bin:/usr/games/bin:/share/darin/bin:/usr/local/kde/bin:/opt/ibm/java2-i386-50/jre/bin:/opt/ibm/java2-i386-50/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @rgs

dmcbride@&#8203;cpan\.org \(via RT\) wrote&#8203;:
> 
> \#\!/usr/bin/perl
> use strict;
> use warnings;
> 
> require List&#8203;::MoreUtils;
> \# use List&#8203;::MoreUtils \(\);
> 
> my @&#8203;v1 = qw\(4 0 0 3\);
> my @&#8203;v2 = qw\(4 0 0 3\);
> 
> my $ea = List&#8203;::MoreUtils&#8203;::each\_array\(@&#8203;v1\, @&#8203;v2\);
> while \(my \($v\, $t\) = $ea\->\(\)\)
> \{
> \}
> \_\_END\_\_
> 
> When running the above\, I get a seg fault\.  Switch out the require to
> use\, and the seg fault goes away\.  This has been confirmed on 5\.8\.7 on
> XP as well in this thread&#8203;: http&#8203;://www\.perlmonks\.org/index\.pl?node\_id=557894

A backtrace gives this :
\#0  0xb7eb6142 in Perl\_av\_fetch \(\) from /usr/lib/perl5/5\.8\.8/i386\-linux/CORE/libperl\.so
\#1  0xb7c531ec in XS\_List\_\_MoreUtils\_\_array\_iterator \(\)
&nbsp;   from /usr/lib/perl5/vendor\_perl/5\.8\.8/i386\-linux/auto/List/MoreUtils/MoreUtils\.so
\#2  0xb7ebd73d in Perl\_pp\_entersub \(\) from /usr/lib/perl5/5\.8\.8/i386\-linux/CORE/libperl\.so
\#3  0xb7eb6ae5 in Perl\_runops\_standard \(\) from /usr/lib/perl5/5\.8\.8/i386\-linux/CORE/libperl\.so
\#4  0xb7e63503 in perl\_run \(\) from /usr/lib/perl5/5\.8\.8/i386\-linux/CORE/libperl\.so
\#5  0x080490d1 in main \(\)

So the problem probably comes from the implementation of
List&#8203;::MoreUtils&#8203;::each\_array \(the \_array\_iterator function being
instantiated as a closure at run\-time\) and is not related to core perl
\(although it's still a bit obscure for me\)

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @smpeters

> \[rgarciasuarez@&#8203;mandriva\.com \- Wed Jun 28 03&#8203;:27&#8203;:53 2006\]&#8203;:
> 
> dmcbride@&#8203;cpan\.org \(via RT\) wrote&#8203;:
> >
> > \#\!/usr/bin/perl
> > use strict;
> > use warnings;
> >
> > require List&#8203;::MoreUtils;
> > \# use List&#8203;::MoreUtils \(\);
> >
> > my @&#8203;v1 = qw\(4 0 0 3\);
> > my @&#8203;v2 = qw\(4 0 0 3\);
> >
> > my $ea = List&#8203;::MoreUtils&#8203;::each\_array\(@&#8203;v1\, @&#8203;v2\);
> > while \(my \($v\, $t\) = $ea\->\(\)\)
> > \{
> > \}
> > \_\_END\_\_
> >
> > When running the above\, I get a seg fault\.  Switch out the require
> to
> > use\, and the seg fault goes away\.  This has been confirmed on 5\.8\.7
> on
> > XP as well in this thread&#8203;:
> http&#8203;://www\.perlmonks\.org/index\.pl?node\_id=557894
> 
> A backtrace gives this :
> \#0  0xb7eb6142 in Perl\_av\_fetch \(\) from /usr/lib/perl5/5\.8\.8/i386\-
> linux/CORE/libperl\.so
> \#1  0xb7c531ec in XS\_List\_\_MoreUtils\_\_array\_iterator \(\)
>    from /usr/lib/perl5/vendor\_perl/5\.8\.8/i386\-
> linux/auto/List/MoreUtils/MoreUtils\.so
> \#2  0xb7ebd73d in Perl\_pp\_entersub \(\) from /usr/lib/perl5/5\.8\.8/i386\-
> linux/CORE/libperl\.so
> \#3  0xb7eb6ae5 in Perl\_runops\_standard \(\) from
> /usr/lib/perl5/5\.8\.8/i386\-linux/CORE/libperl\.so
> \#4  0xb7e63503 in perl\_run \(\) from /usr/lib/perl5/5\.8\.8/i386\-
> linux/CORE/libperl\.so
> \#5  0x080490d1 in main \(\)
> 
> So the problem probably comes from the implementation of
> List&#8203;::MoreUtils&#8203;::each\_array \(the \_array\_iterator function being
> instantiated as a closure at run\-time\) and is not related to core perl
> \(although it's still a bit obscure for me\)
> 

This ticket has been closed and CPAN ticket \#20258&#8203;: Core dump in
each\_array iterator has been opened
\<http&#8203;://rt\.cpan\.org/Public/Bug/Display\.html?id=20258>\.

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'rejected'
